### PR TITLE
Add Universe Selection Tool

### PR DESCRIPTION
**Pull request by LouisSzeto**
- Allow universe selection by `QuiverQuantTwitterFollowersUniverse` class
- Add DataProcessing Tools to create universe files (C#/Py)
- Add Unit Tests

### CODE REVIEW

File: DataProcessing/QuantConnect.DataProcessing.csproj [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    <PropertyGroup>
+        <OutputType>Exe</OutputType>
+        <TargetFramework>net5.0</TargetFramework>
+        <AssemblyName>process</AssemblyName>
+        <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
+    </PropertyGroup>
+    <ItemGroup>
+        <PackageReference Include="QuantConnect.Lean.Engine" Version="2.5.*"/>
+        <PackageReference Include="QuantConnect.Research" Version="2.5.*"/>
+        <PackageReference Include="QuantConnect.pythonnet" Version="2.0.11" />
```
Comment (Martin-Molinero) [inline DataProcessing/QuantConnect.DataProcessing.csproj]: Might be a good idea to use here too `2.0.*` 👌

File: DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+                .ToList();
+
+            var tempPath = Path.Combine(Path.GetTempPath(), $"{Guid.NewGuid()}.tmp");
+            File.WriteAllLines(tempPath, finalLines);
+            var tempFilePath = new FileInfo(tempPath);
+            if (finalFileExists)
+            {
+                tempFilePath.Replace(finalPath, bkPath);
+                var bkFilePath = new FileInfo(bkPath);
+                bkFilePath.Delete();
```
Comment (Martin-Molinero) [inline DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs]: Why use a backup just to delete it right away, I think we can directly skip creating the backup file?

File: DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+
+                                        var universeCsvContents = $"{sid},{ticker},{follower},{dayChange},{weekChange},{monthChange}";
+
+                                        _tempData.GetOrAdd(date, new ConcurrentQueue<string>());
+
+                                        if (_tempData.TryGetValue(date, out var tempList))
+                                        {
+                                            tempList.Enqueue(universeCsvContents);
+                                        }
+                                        _tempData.TryUpdate(date, tempList, _tempData[date]);
```
Comment (Martin-Molinero) [inline DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs]: `GetOrAdd` returns the value don't need to call `TryGetValue` right after? Don't think `TryUpdate` would be required either?

File: DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+
+                    // Makes sure we don't overrun Quiver rate limits accidentally
+                    _indexGate.WaitToProceed();
+
+                    tasks.Add(
+                        HttpRequester($"historical/{VendorDataName}/{ticker}")
+                            .ContinueWith(
+                                y =>
+                                {
+                                    i++;
```
Comment (Martin-Molinero) [inline DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs]: > i++;

This isn't thread safe, should use `Interlocked.Increment(ref i);` instead

File: DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+                                        SaveContentToFile(_destinationFolder, ticker, csvContents);
+                                    }
+
+                                    var percentageDone = i / count;
+                                    if (percentageDone >= currentPercent)
+                                    {
+                                        Log.Trace(
+                                            $"QuiverQuantTwitterFollowersDataDownloader.Run(): {percentageDone.ToStringInvariant("P2")} complete");
+                                        currentPercent += percent;
+                                    }
```
Comment (Martin-Molinero) [inline DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs]: Think this section isn't thread safe either? should probably use the returned value of `Interlocked.Increment` above to determine if should log, like:
```C#
                                    var newValue = Interlocked.Increment(ref i);
                                    if (newValue % 100 == 0)
```

File: DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+    /// </summary>
+    public class QuiverQuantTwitterFollowersDataDownloader : IDisposable
+    {
+        public const string VendorName = "quiver";
+        public const string VendorDataName = "twitter";
+        
+        private readonly string _destinationFolder;
+        private readonly string _universeFolder;
+        private readonly string _clientKey;
+        private readonly string _dataFolder = Config.Get("data-folder", Globals.DataFolder);
```
Comment (Martin-Molinero) [inline DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs]: Suggest using `Globals.DataFolder` directly which is further abstracted from the data folder config value really

File: DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+                        Log.Error(
+                            $"QuiverQuantTwitterFollowersDataDownloader(): Defunct ticker {quiverTicker} is unable to be parsed. Continuing...");
+                        continue;
+                    }
+
+                    // Begin processing ticker with a normalized value
+                    Log.Trace($"QuiverQuantTwitterFollowersDataDownloader.Run(): Processing {ticker}");
+
+                    // Makes sure we don't overrun Quiver rate limits accidentally
+                    _indexGate.WaitToProceed();
```
Comment (Martin-Molinero) [inline DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs]: Think rate gate wait should be inside `HttpRequester`, closer where we actually make the request, else for example it wouldn't cover retries?

File: DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+            return true;
+        }
+
+
+        /// <summary>
+        /// Normalizes Estimize tickers to a format usable by the <see cref="Data.Auxiliary.MapFileResolver"/>
+        /// </summary>
+        /// <param name="ticker">Ticker to normalize</param>
+        /// <returns>Normalized ticker</returns>
+        private static string NormalizeTicker(string ticker)
```
Comment (Martin-Molinero) [inline DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs]: nit - unused method, should be used/removed?

File: DataProcessing/process.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+# 4. Output data for each ticker to /temp-output-directory/alternative/quiver/twitter/{symbol}.csv
+
+import os
+from clr_loader import get_coreclr
+from pythonnet import set_runtime
+set_runtime(get_coreclr(os.path.join(os.path.dirname(os.path.realpath(__file__)), "config.json")))
+
+from AlgorithmImports import *
+from QuantConnect.Lean.Engine.DataFeeds import *
+AddReference("Fasterflect")
```
Comment (Martin-Molinero) [inline DataProcessing/process.py]: Think we should push this section to it's own file in the template/sdk repo and just reuse it when required. I think ideally we should remove the dependency of the `config.json`. We can probably set the runtime providing it as a string versus the lib reading it from the file

File: DataProcessing/process.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+        companies = sorted(self.HttpRequester("companies"), key=lambda x: x['Ticker'])
+    
+        for c in companies:
+            ticker = c['Ticker']
+            filename = os.path.join(self.destinationFolder, f'{ticker.lower()}.csv')
+            print(f'Processing ticker: {ticker}')
+            i = 5
+
+            while i != 0:
```
Comment (Martin-Molinero) [inline DataProcessing/process.py]: Think we should expand variable names `i`,`c`, `dt`, `fp` etc to be more representative when possible

File: DataProcessing/process.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+                try:
+                    ticker_twitter = self.HttpRequester(f"historical/{VendorDataName}/{ticker}")
+                    time.sleep(0.03)
```
Comment (Martin-Molinero) [inline DataProcessing/process.py]: > time.sleep(0.03)

Why is this required, rate gate? Is there some more normalized rate gate we could use for python?

File: DemostrationUniverseSelection.cs [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+                       where d.Followers > 200000 && d.WeekPercentChange > 0m
+                       select d.Symbol;
+            });
+        }
+        
+        public override void OnSecuritiesChanged(SecurityChanges changes)
+        {
+        	Log(changes.ToString());
+        }
+    }
```
Comment (Martin-Molinero) [inline DemostrationUniverseSelection.cs]: Cool! indentation is off in a couple of places, would be nice to add a description to the class too. I think ideally we should also add a python version of the algorithm 👍

File: DataProcessing/Program.cs [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+                if (!success)
+                {
+                    Log.Error($"QuantConnect.DataProcessing.Program.Main(): Failed to download/process {QuiverQuantTwitterFollowersDataDownloader.VendorName} {QuiverQuantTwitterFollowersDataDownloader.VendorDataName} data");
+                    return 1;
+                }
+            }
+            catch (Exception err)
+            {
+                Log.Error(err, $"The downloader/converter for {QuiverQuantTwitterFollowersDataDownloader.VendorDataName} {QuiverQuantTwitterFollowersDataDownloader.VendorDataName} data exited unexpectedly");
+                return 1;
```
Comment (Martin-Molinero) [inline DataProcessing/Program.cs]: Nice!
I think might be better to use `Environment.Exit()` instead of `return` just to avoid any weird hanging issues. 
nit - couple of logs not following logging style -> missing `QuantConnect.DataProcessing.Program.Main()`

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Nice! leaving a few requests

File: DataProcessing/process.py (hunk 11 above)
Comment (LouisSzeto) [reply DataProcessing/process.py]: Seems no...

File: DataProcessing/Program.cs [hunk 14]
```diff
@@ -47,8 +47,8 @@ public static int Main()
             }
             catch (Exception err)
             {
-                Log.Error(err, $"The downloader/converter for {QuiverQuantTwitterFollowersDataDownloader.VendorDataName} {QuiverQuantTwitterFollowersDataDownloader.VendorDataName} data failed to be constructed");
-                return 1;
+                Log.Error(err, $"QuantConnect.DataProcessing.Program.Main(): The downloader/converter for {QuiverQuantTwitterFollowersDataDownloader.VendorDataName} {QuiverQuantTwitterFollowersDataDownloader.VendorDataName} data failed to be constructed");
+                return;
```
Comment (Martin-Molinero) [inline DataProcessing/Program.cs]: Should be `Environment.Exit(1);` ?

File: DataProcessing/CLRImports.py [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+                 }
+             }
+         },
+        fp, ensure_ascii=True, indent=4)
+set_runtime(get_coreclr('tmp.json'))
+remove('tmp.json')
+
+from AlgorithmImports import *
+from QuantConnect.Lean.Engine.DataFeeds import *
+AddReference("Fasterflect")
```
Comment (Martin-Molinero) [inline DataProcessing/CLRImports.py]: Cool! Would be nice to add some comments, what is this, when/how/why should we use it

File: DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs [hunk 16]
```diff
@@ -345,9 +329,7 @@ private void SaveContentToFile(string destinationFolder, string name, IEnumerabl
             var tempFilePath = new FileInfo(tempPath);
             if (finalFileExists)
             {
-                tempFilePath.Replace(finalPath, bkPath);
-                var bkFilePath = new FileInfo(bkPath);
-                bkFilePath.Delete();
+                tempFilePath.Replace(finalPath, null);
```
Comment (Martin-Molinero) [inline DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs]: Can use `MoveTo` directly?

File: DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs [hunk 17]
```diff
@@ -177,25 +165,19 @@ public bool Run()
                                         var universeCsvContents = $"{sid},{ticker},{follower},{dayChange},{weekChange},{monthChange}";
 
                                         _tempData.GetOrAdd(date, new ConcurrentQueue<string>());
-
-                                        if (_tempData.TryGetValue(date, out var tempList))
-                                        {
-                                            tempList.Enqueue(universeCsvContents);
-                                        }
-                                        _tempData.TryUpdate(date, tempList, _tempData[date]);
+                                        _tempData[date].Enqueue(universeCsvContents);
```
Comment (Martin-Molinero) [inline DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs]: > _tempData[date]

I think could directly use the returned value of `GetOrAdd` above? avoiding the double access

File: DemostrationUniverseSelection.cs (hunk 12 above)
Comment (Martin-Molinero) [reply DemostrationUniverseSelection.cs]: Would be nice to add the python demonstration version too

File: DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs (hunk 6 above)
Comment (Martin-Molinero) [reply DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs]: Ping

**Review by Martin-Molinero — COMMENTED** [review]
Cool! looking nice, leaving minor final comments

File: DataProcessing/Program.cs (hunk 14 above)
Comment (LouisSzeto) [reply DataProcessing/Program.cs]: `"instance" is not defined` when build (line 59)

File: QuiverQuantTwitterFollowers.cs [hunk 18]
```diff
@@ -169,7 +185,7 @@ public override List<Resolution> SupportedResolutions()
         {
             return DailyResolution;
         }
-
+
```
Comment (AlexCatarino) [inline QuiverQuantTwitterFollowers.cs]: We should remove this empty tab/spaces.

File: QuiverQuantTwitterFollowersUniverse.cs [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+            );
+        }
+
+        public override BaseData Reader(SubscriptionDataConfig config, string line, DateTime date, bool isLiveMode)
+        {
+            var csv = line.Split(',');
+
+            return new QuiverQuantTwitterFollowersUniverse
+            {
+                Followers = Parse.Int(csv[2]),
```
Comment (AlexCatarino) [inline QuiverQuantTwitterFollowersUniverse.cs]: I think we should set the `Base.Value` to something. In the other class, I set it to `Followers`.

File: QuiverQuantTwitterFollowersUniverseSelectionAlgorithm.cs [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+	    SetStartDate(2022, 2, 14);
+            SetEndDate(2022, 2, 18);
+            SetCash(100000);
+
+            // add a custom universe data source (defaults to usa-equity)
+            AddUniverse<QuiverQuantTwitterFollowersUniverse>("QuiverQuantTwitterFollowersUniverse", Resolution.Daily, data =>
+            {
+            	foreach (var datum in data)
+            	{
+            		Log($"{datum.Symbol},{datum.Followers},{datum.DayPercentChange},{datum.WeekPercentChange}");
```
Comment (AlexCatarino) [inline QuiverQuantTwitterFollowersUniverseSelectionAlgorithm.cs]: nit. In this line, the indentation is not correct.

File: QuiverQuantTwitterFollowersUniverse.cs [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+                DayPercentChange = decimal.Parse(csv[3], NumberStyles.Any, CultureInfo.InvariantCulture),
+                WeekPercentChange = decimal.Parse(csv[4], NumberStyles.Any, CultureInfo.InvariantCulture),
+                MonthPercentChange = decimal.Parse(csv[5], NumberStyles.Any, CultureInfo.InvariantCulture),
+
+                Symbol = new Symbol(SecurityIdentifier.Parse(csv[0]), csv[1]),
+                Time = date.AddDays(-1),
```
Comment (AlexCatarino) [inline QuiverQuantTwitterFollowersUniverse.cs]: In the other class, we are not subtracting one day from `Time`. Are the timestamps and values of the Universe and Daily file matching?

File: QuiverQuantTwitterFollowersUniverse.cs [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+        /// Week-over-week change in company's follower count
+        /// </summary>
+        public decimal WeekPercentChange { get; set; }
+
+        /// <summary>
+        /// Month-over-month change in company's follower count
+        /// </summary>
+        public decimal MonthPercentChange { get; set; }
+
+        public override DateTime EndTime
```
Comment (AlexCatarino) [inline QuiverQuantTwitterFollowersUniverse.cs]: In the other class, we have
```csharp
public override DateTime EndTime => Time + Period;
```
and define `Period` as
```csharp
public TimeSpan Period { get; set; } = TimeSpan.FromDays(1);
```

**Review by AlexCatarino — CHANGES_REQUESTED** [review]
Thank you @LouisSzeto.
I left some comments about standardization between the single asset and universe classes.

File: DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs (hunk 16 above)
Comment (Martin-Molinero) [reply DataProcessing/QuiverQuantTwitterFollowersDataDownloader.cs]: `MoveTo` has an option to override, can remove the need for this `if/else`

File: QuiverQuantTwitterFollowersUniverse.cs [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+        /// <summary>
+        /// Month-over-month change in company's follower count
+        /// </summary>
+        public decimal MonthPercentChange { get; set; }
+
+        public TimeSpan Period { get; set; } = TimeSpan.FromDays(1);
+
+        public override DateTime EndTime => Time + Period;
+        
+        public override SubscriptionDataSource GetSource(SubscriptionDataConfig config, DateTime date, bool isLiveMode)
```
Comment (Martin-Molinero) [inline QuiverQuantTwitterFollowersUniverse.cs]: nit - Missing xml docs

File: QuiverQuantTwitterFollowersUniverseSelectionAlgorithm.cs [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+namespace QuantConnect.Algorithm.CSharp
+{
+    public class QuiverQuantTwitterFollowerUniverseAlgorithm : QCAlgorithm
+    {
+        public override void Initialize()
+        {
+            // Data ADDED via universe selection is added with Daily resolution.
+            UniverseSettings.Resolution = Resolution.Daily;
+
+	    SetStartDate(2022, 2, 14);
```
Comment (Martin-Molinero) [inline QuiverQuantTwitterFollowersUniverseSelectionAlgorithm.cs]: ping

File: QuiverQuantTwitterFollowersUniverseSelectionAlgorithm.cs [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+            SetEndDate(2022, 2, 18);
+            SetCash(100000);
+
+            // add a custom universe data source (defaults to usa-equity)
+            AddUniverse<QuiverQuantTwitterFollowersUniverse>("QuiverQuantTwitterFollowersUniverse", Resolution.Daily, data =>
+            {
+            	foreach (var datum in data)
```
Comment (Martin-Molinero) [inline QuiverQuantTwitterFollowersUniverseSelectionAlgorithm.cs]: Indentation is wrong here too?
